### PR TITLE
Draft: Fix stdClass cast

### PR DESCRIPTION
**Pull request by edudobay**
https://github.com/vimeo/psalm/issues/8187

I followed @orklah's suggestion of typing object casts as `stdClass&object{key:value}`, but my solution broke other tests. Turns out that `stdClass` seems to always "test positive" for existence of any property.

For starters, I added a silly check for the specific shape `stdClass & object{some properties here}`, but that doesn't seem right. I'd appreciate suggestions for improving that!

### CODE REVIEW

File: src/Psalm/Internal/Provider/ReturnTypeProvider/GetObjectVarsReturnTypeProvider.php [hunk 1]
```diff
@@ -41,6 +42,14 @@ public static function getGetObjectVarsReturnType(
             $atomics = $first_arg_type->getAtomicTypes();
             $object_type = reset($atomics);
 
+            if ($object_type instanceof TNamedObject
+                && strtolower($object_type->value) === strtolower(stdClass::class)
+                && count($intersection_types = $object_type->getIntersectionTypes() ?: []) === 1
+                && ($inner_object_type = reset($intersection_types)) instanceof TObjectWithProperties
+            ) {
+                $object_type = $inner_object_type;
+            }
```
Comment (AndrolGenhald) [inline src/Psalm/Internal/Provider/ReturnTypeProvider/GetObjectVarsReturnTypeProvider.php:51]: This seems a bit hacky, I think what we should probably do is change how `Config::$universal_object_crates` works so that unknown properties aren't allowed to be read without checking if they exist, but writing them is fine (and should produce `&object{newProperty: [type]`, eg for `$obj = object{foo: string, bar: string}`, `$obj->foobar = 1` results in `object{foo: string, bar: string, foobar: 1}`).

Basically I'd like to see [this](https://github.com/vimeo/psalm/issues/7916#issuecomment-1115435524) fixed (so that [this](https://psalm.dev/r/75644eb2ad) can't happen), but maybe there's something I'm not considering here, @orklah do you know if there's a good reason it works like that?

Comment (AndrolGenhald) [reply src/Psalm/Internal/Provider/ReturnTypeProvider/GetObjectVarsReturnTypeProvider.php:51]: Also, since this would be a bigger change, if we go with it maybe this should target master?

Comment (orklah) [reply src/Psalm/Internal/Provider/ReturnTypeProvider/GetObjectVarsReturnTypeProvider.php:51]: I don't think there is a big design decision behind how it works right now. Universal object crates, intersection types (on named object) and the cast from array to object were just three separate modules that happen to work relatively well with each other

Comment (edudobay) [reply src/Psalm/Internal/Provider/ReturnTypeProvider/GetObjectVarsReturnTypeProvider.php:51]: Seems fair! I'll try to figure out something about this on the next few days :)

**Comment by weirdan** [comment]
4.x branch is closed now as we prepare for the 5.0 release. Please target the master branch instead.

**Comment by orklah** [comment]
Closing as this targets 4.x which is closed. Please reopen on master going forward
